### DISCUSSION TITLE
How to initiate connections behind the same NAT router

### DISCUSSION THREAD

**Discussion by andrew-hoff** [strongswan/strongswan#431] [Q&A]
I have a setup with three nodes (A, B and C) all of which are behind NAT routers. In this case, A and B both connect to C, but not to each other. Further, A and B both initiate their connections with C. I load the initial connection on each device using vici with the following configuration:

`Version: 2
MobIKE: "yes"
DeadPeerDetectionDelay: 10
Proposals: <proposals>
Local:
    Auth: <auth>
    ID: <connection ID>
Remote:
    Auth: <auth>
    ID:  "%any"
LocalAddrs: <IP address of the local node>
RemotePort: <custom NAT-T port i.e. I use a non-4500 port>
Unique: "no"
Reauth Time: <Reauth time>
Rekey Time: <Rekey time>
Keying Tries: <Keying tries>
Children:
    Mode: "transport"
    LocalTrafficSelectors: {0.0.0.0/0[gre/<key>], ::/0[gre/<key>]}
    RemoteTrafficSelectors: {0.0.0.0/0[gre/<key>], ::/0[gre/<key>]}
    ESPProposals: <proposals>
    DeadPeerDetectionAction: "restart"
`

On nodes A and B, I also fill in the `RemoteAddrs` field, but I do _not_ do this for node C.

I noticed that when I connect only A to C, everything works just fine. However, when I connect B as well, I start getting `TS_UNACCEPTABLE` errors in my logs for both A and B. To mitigate this, I added the `RemoteAddrs` field to C. This fixed my problem, which makes sense: prior to this change strongswan had no way to distinguish between the two loaded connections on node C.

However, what happens if both A and B are behind the _same_ NAT router? They'll have the same remote address and will then have the indistinguishable connections, leaving me in the same place I started. Are there other fields I should set to distinguish between the two when using NAT?

**Comment by tobiasbrunner** [comment]
Maybe assign virtual IPs? (Since you seem to be doing something with GRE that might get complicated, but so will probably any workaround for this - may be easier to make sure that there are no conflicts or get rid of the NATs entirely e.g. by using IPv6.)

**Reply by andrew-hoff** [reply]
But doesn't the VIP only get applied _after_ the connection is initiated? In other words, if I initiate a new connection, I don't have a VIP yet, so it can't be used to distinguish which connection to initiate.

Technically, I already have a VIP on the GRE interface, but to use that, I'd need to know the peers VIP before initiating the connection.

**Reply by tobiasbrunner** [reply]
> But doesn't the VIP only get applied _after_ the connection is initiated?

If it's negotiated, yes. You could also assign them statically. Or use some kind of NAT to map the stuff.



>  I don't have a VIP yet, so it can't be used to distinguish which connection to initiate.

No sure what you mean by that.

**Reply by andrew-hoff** [reply]
Ah, to clarify - I can and do assign VIPs statically to each GRE interface. However, the peers are not aware of each other's VIP until _after_ the connection is established because I don't want to transmit the VIP in plaintext. As a result, if A loads a connection for B, it won't know B's VIP until an SA has been established.
